cfep-22.md

```diff
@@ -0,0 +1,185 @@
+
+Recipes whose package version is the same as the ABI version do not need to
+modify their build string.
+
+Recipes with multiple library artifacts whose ABIs are tracked separately
+should split these libraries into separate outputs of one recipe. A metapackage
+may be used for installation convenience, but the metapackge must also
+enumerate the run_exports for each subpackage.
+
+## Sample Implementation
+
+```yaml
+{% set name = "libavif" %}
+{% set build = 0 %}
+# NOTE: Humans must also update the library version in the tests section of this recipe
+{% set version = "0.10.1" %}
+# Look in the libavif top level CMakeLists.txt for the updated ABI version.
+# ABI is updated separately from API version.
+{% set so_version = "14.0.1" %}
+{% set so_major_version = so_version.split('.')[0] %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+build:
+  number: {{ build }}
+  string: "v{{ so_name_major }}soh{{ PKG_HASH }}_{{ build }}"
+  run_exports:
+    - {{ pin_subpackage(name|lower) }} *v{{ so_name_major }}so*
+
+test:
+  commands:
+    - test -f ${PREFIX}/lib/libavif.so.{{ so_major_version }}     # [linux]
+    - test -f ${PREFIX}/lib/libavif.so.{{ so_version }}           # [linux]
+
+```
+
+## Rationale
+
+Adding the ABI version to the build string allows the run_export to handle both
+API and ABI using existing conda capabilities. Not needing new conda features
+saves engineering effort and is implementable today. Build strings are already
+used to track features such as the CUDA version and MPI variants.
+
+The ABI version is sandwiched between `v` and `so` because these letters are
+not part of hexadecimal and to prevent matching collisions with other build
+string tags or higher numbers. i.e. `6*` matches with `60`, but `6so*` does
+not.
+
+Using a combination of build string and run_exports to add ABI information
+requires no changes from downstream packages except using the latest build.
+They will be guarded from ABI breakages as soon as they rebuild with the
+updated package.
+
+## Backward Compatability
+
+This proposal does not break exisiting packages. It is a feature that only
+helps future package builds.
+
+
+## Alternatives
+
+### API Pinning Only
+
+In theory, breaking ABI changes can be introduced without changing the API
+(reordering struct elements for example). In practice, project managers would
+always make a new API release for this change, so ABI breaks can be avoided by
+pinning down to the patch version. This is the most conservative approach, but
+is the least flexible. Recipe maintainers can pin to minor API versions if the
+upstream package makes any such promises about not breaking the ABI.
+
+### Adding ABI name to package name
+
+This alternatives renames outputs to `{{ name|lower }}{{ so_major_version }}`.
+This is not backward compatible and would require migrating all downstream
+packages to the new output names. However, it does enable installing multiple
```

Comment:
This is not true. There's a `{{ name|lower }}` output too, so it is backwards compatible. Please see

Reply:
That's a good point! I'll update this section with a better description and example using your suggestion.

Reply:
I have added a more detailed example of this alternative. I think it is the best alternative. I assume that its major utility is that it allows for multiple ABIs of a library to be installed into the same environment. Judging whether this is an important feature for conda-forge is beyond my depth.

I prefer the build string approach because it is simpler and because I haven't run into a case when I wanted to install multiple versions of the same library. However, this is just my personal experience.

Reply:
Allowing multiple ABI versions of a package simultaneously causes much more work for refactoring a recipe in order to prevent clobbering. I have updated the proposal to better show how splitting a recipe this way would work.

Reply:
Related thought: The purpose of the global conda-forge-pinnings/migrations is to keep all packages in sync with respect to the ABI/API of popular packages. Since publishing ABI-in-the-name packages allows installing multiple ABIs of the same package, does that mean that global conda-forge-pinnings would no longer be needed?

If so, which is easier? Maintaining global pinnings or maintaining individual feedstocks to ensure that ABI-in-the-name packages are implemented (correctly).

Reply:
I'm not suggesting we do multiple ABIs at the same time. We have both `libgfortran` and `libgfortran5` as `run_exports` and if a person really wants multiple ABIs, they can do that, but that's not the default. Because both `libgfortran` and `libgfortran5` are dependencies, you can only have one package.